File: packages/swift/Sources/EcollectSDK/Utils/Polling.swift
import Foundation

/// Terminal transaction states — polling stops when one of these is received.
public let finalTranStates: Set<String> = ["OK", "NOT_AUTHORIZED", "EXPIRED", "FAILED"]

/// Intermediate states for which the SDK will NOT automatically retry a payment
/// (double payment prevention). Only poll for status updates in these states.
public let doublePaymentProtectedStates: Set<String> = ["BANK", "PENDING", "CAPTURED", "CREATED"]

/// Manages polling for transaction resolution using Swift Concurrency.
public final class PollingManager {

    /// Polls getTransactionInformation until a final state is reached or timeout expires.
    ///
    /// - Parameters:
    ///   - ticketId: The ecollect TicketId to poll.
    ///   - timeout: Maximum duration in seconds (default 600 = 10 minutes).
    ///   - fetch: Async closure that fetches the current `TransactionInfoResponse`.
    /// - Returns: The `TransactionInfoResponse` with a final `TranState`.
    /// - Throws: `EcollectError.pollingTimeout` if no final state is reached in time.
    public static func poll(
        ticketId: Int,
        timeout: TimeInterval = 600,
        fetch: @escaping () async throws -> TransactionInfoResponse
    ) async throws -> TransactionInfoResponse {
        let deadline = Date().addingTimeInterval(timeout)

        while true {
            let status = try await fetch()
            let state = status.TranState ?? ""

            if finalTranStates.contains(state) {
                return status
            }

            if Date() >= deadline {
                throw EcollectError.pollingTimeout(
                    "Polling timed out after \(Int(timeout))s for TicketId \(ticketId). Last state: \(state)"
                )
            }

            // BANK / PENDING: poll every 30 seconds
            // CREATED: user may have abandoned, poll every 5 minutes
            let interval: UInt64
            if state == "CREATED" {
                interval = 300_000_000_000 // 5 minutes
            } else {
                interval = 30_000_000_000  // 30 seconds
            }
            try await Task.sleep(nanoseconds: interval)
        }
    }
}

File: packages/python/ecollect/utils/polling.py
"""Async polling manager for transaction state resolution."""
import asyncio
import logging
from typing import Any, Callable, Optional

from ecollect.exceptions import PollingTimeoutError

logger = logging.getLogger(__name__)

_TERMINAL_STATES = {"OK", "NOT_AUTHORIZED", "EXPIRED", "FAILED"}
_BANK_PENDING_INTERVAL = 30  # seconds
_CREATED_INTERVAL = 300  # seconds


class PollingManager:
    """Poll getTransactionInformation until a terminal state is reached."""

    async def start_polling(
        self,
        ticket_id: int,
        get_status: Callable[[int], Any],
        on_complete: Callable[[Any], None],
        timeout: int = 600,
    ) -> None:
        """Poll transaction status until terminal state or timeout.

        Args:
            ticket_id: ecollect TicketId to poll.
            get_status: async callable that accepts ticket_id and returns
                        a dict with at minimum a 'TranState' key.
            on_complete: callback invoked with the final status dict.
            timeout: maximum polling duration in seconds (default 600s / 10 min).
        """
        elapsed = 0.0
        while elapsed < timeout:
            status = await get_status(ticket_id)
            tran_state = status.get("TranState", "")

            if tran_state in _TERMINAL_STATES:
                on_complete(status)
                return

            if tran_state in ("BANK", "PENDING"):
                interval = _BANK_PENDING_INTERVAL
            elif tran_state == "CREATED":
                interval = _CREATED_INTERVAL
            else:
                interval = _BANK_PENDING_INTERVAL

            remaining = timeout - elapsed
            wait = min(interval, remaining)
            if wait <= 0:
                break

            logger.debug(
                "TicketId=%s state=%s, polling again in %ds", ticket_id, tran_state, wait
            )
            await asyncio.sleep(wait)
            elapsed += wait

        raise PollingTimeoutError(
            f"Polling timed out after {timeout}s for TicketId={ticket_id}"
        )
